### PR TITLE
Story 2 - Edit applicants

### PR DESCRIPTION
**Pull request by fennel3**
As an: Academy Admin

I want: To be able to edit applicants

So that: I can correct mistakes made when adding an applicant, or when an applicant changes their details

We have assigned this 8 story points.

Requirements:

Must display an ‘Edit’ button next to each applicant on the applicants page in the ‘Actions’ column

This should take you to a seperate edit page

Must be able to edit all applicant and application fields

All form fields should be pre-populated with the existing data for that applicant

If the applicant does not have a row in the applications table, do not display any of the application fields on the edit page

Should be redirected to the single applicant page for the edited applicant upon success

Must display appropriate validation messages (should be consistent with adding an applicant)

### CODE REVIEW

File: templates/editApplicant.phtml
Comment (mallowmew) [inline templates/editApplicant.phtml:1]: Confusingly, this page looks different from the Add Applicant page? These pages should look visually similar - because they're basically the same page with a different heading.

Comment (mallowmew) [inline templates/editApplicant.phtml:1]: The fields in the form should be in the same order as on the add applicant page.

The label text should be the same as the Add Applicant page.

Basically copy paste the entire form from that page witht he same CSS, and only put in the PHP stuff that's actually different here.

(Note: unfortunately you can't use a partial template because there are differences in the PHP 😢 - unless you do something clever?)

File: templates/applicants.phtml [hunk 1]
```diff
@@ -42,6 +43,7 @@
                     <td><a href="/admin/applicants/<?= $applicant->getId() ?>"><?= $applicant->getName(); ?></a></td>
                     <td><?php echo $applicant->getEmail(); ?></td>
                     <td><?php echo $applicant->getFormattedApplicationDate(); ?></td>
+                    <td><a href="/admin/applicants/edit/<?= $applicant->getId() ?>" class="edit-button">EDIT</a></td>
```
Comment (mallowmew) [inline templates/applicants.phtml]: Button should look like other buttons in the application. The colour can be different (and I do like the choice of the same pink as the iO Academy logo) but the border, corner-roundness, padding, etc. should be consistent.

This is a business application, so **no fun allowed**. The button should not have the pop-out hover effect. A hover color change with or without a CSS transition is fine, but make it transition to that light grey colour the other button goes to because the purple you chose is not in the colour scheme. We should be cautious adding colours that aren't in the colour scheme given to us by a designer because that's not really our role and it's the sort of thing that tends to become stuck.

File: public/css/style.css
Comment (mallowmew) [inline public/css/style.css:1]: Consider using CSS variables for the colour scheme. Now that we have multiple buttons and stuff it might get annoying to need to remember hex codes.

Maybe check if the SASS files have a color theme defined as variables that you can just copy?

File: src/Controllers/Controller.php [hunk 2]
```diff
@@ -10,7 +10,7 @@
  */
 abstract class Controller
 {
-    abstract public function __invoke(Request $request, Response $response): Response;
+    abstract public function __invoke(Request $request, Response $response, $args): Response;
```
Comment (mallowmew) [inline src/Controllers/Controller.php]: You can add this with a default empty array so that you don't need to edit all of the controllers - I was asked about something like this earlier so I must've misunderstood, sorry!

```suggestion
    abstract public function __invoke(Request $request, Response $response, $args = []): Response;
```

File: src/Models/ApplicantsModel.php [hunk 3]
```diff
@@ -97,6 +105,50 @@ public function getById(int $id): Applicant|false
...
+            'cohort_id' => $details['cohort_id'],
+            'dob' => $details['dob'],
+            'phone' => $details['phone'],
+            'address' => $details['address'],
+            'heard_about_id' => $details['heard_about_id'],
+            'age_confirmation' => $details['age_confirmation'],
+            'newsletter' => $details['newsletter'],
+            'eligible' => $details['eligible'],
+            'terms' => $details['terms']
+        ]);
```
Comment (mallowmew) [inline src/Models/ApplicantsModel.php]: This changes an Application in the database, so should be in a function in the application model.

It's going to be very similar to the controller for the add applicant page.

File: src/Models/ApplicantsModel.php [hunk 4]
```diff
@@ -44,6 +44,13 @@ public function getCount(): int
         return $query->fetch()['count'];
     }
 
+    public function getCircumstanceID()
+    {
+        $query = $this->db->prepare("SELECT `circumstance_id` FROM `applications`;");
+        $query->execute();
+        return $query->fetch();
+    }
```
Comment (mallowmew) [inline src/Models/ApplicantsModel.php]: Seems to duplicate something you did in story 1?

File: src/Entities/Application.php [hunk 5]
```diff
@@ -87,7 +90,7 @@ public function getExperience(): string
         return $this->experience;
     }
 
-    public function getDiversitech(): bool
+    public function getDiversitech(): int
```
Comment (mallowmew) [inline src/Entities/Application.php]: The `bool`s should stay `bool`, revert this change.

File: src/Entities/Applicant.php [hunk 6]
```diff
@@ -12,18 +12,24 @@ class Applicant
...
-        string $name,
+        int          $id,
+        string       $name,
         EmailAddress $email,
-        string $application_date,
+        string       $application_date,
+
         ?Application $application = null
-    ) {
+
```
Comment (mallowmew) [inline src/Entities/Applicant.php]: Don't format like this unless CircleCI tells you to - inconsistent with the rest of the app.

File: src/Controllers/Pages/EditApplicationController.php [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    private ApplicantsModel $applicantsModel;
+    private CohortsModel $cohortsModel;
+    private AuthService $authService;
+
+
+    public function __construct(ApplicantsModel $model,
+                                PhpRenderer $view,
+                                AuthService $authService,
+                                ApplicantsModel $applicantsModel,
+                                CohortsModel $cohortsModel
```
Comment (mallowmew) [inline src/Controllers/Pages/EditApplicationController.php]: Pretty sure the rest of the application is formatted this way?

```suggestion
    public function __construct(
        ApplicantsModel $model,
        PhpRenderer $view,
        AuthService $authService,
        ApplicantsModel $applicantsModel,
        CohortsModel $cohortsModel
```

File: src/Controllers/FormActions/EditApplicantActionController.php [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+        $details = $request->getParsedBody();
+        $details['id'] = $args['id'];
+
+//        try {
+//            ApplicantValidator::validate($details);
+//            ApplicationValidator::validate($details);
+//        } catch (Exception $e) {
+//            return $this->redirectWithError($response, '/admin/applicant/edit/'.$details['id'], $e->getMessage());
+//        }
```
Comment (mallowmew) [inline src/Controllers/FormActions/EditApplicantActionController.php]: No validation (but you already did it in story 1)

**Review by mallowmew — CHANGES_REQUESTED** [review]

File: src/Models/CohortsModel.php [hunk 9]
```diff
@@ -53,4 +53,12 @@ public function getAll(): array
...
         return $cohorts;
     }
+
+    public function getDates(): array
+    {
+        $query = $this->db->prepare("SELECT `id`, `date` FROM `cohorts`;");
+
+        $query->execute();
+        return $query->fetchAll();
+    }
```
Comment (mallowmew) [inline src/Models/CohortsModel.php]: Seems really similar to something you did on story 1?

File: templates/editApplicant.phtml
Comment (Liamslost) [reply templates/editApplicant.phtml:1]: on it

File: templates/applicants.phtml (hunk 1 above)
Comment (FredJRice) [reply templates/applicants.phtml]: Fred will do this issue

File: templates/editApplicant.phtml
Comment (Liamslost) [reply templates/editApplicant.phtml:1]: on it

File: src/Controllers/Pages/EditApplicationController.php (hunk 7 above)
Comment (vincedefreitas) [reply src/Controllers/Pages/EditApplicationController.php]: Done

File: src/Entities/Applicant.php (hunk 6 above)
Comment (vincedefreitas) [reply src/Entities/Applicant.php]: Done

File: src/Models/CohortsModel.php (hunk 9 above)
Comment (fennel3) [reply src/Models/CohortsModel.php]: Ill take this

File: src/Entities/Application.php (hunk 5 above)
Comment (vincedefreitas) [reply src/Entities/Application.php]: Done

File: src/Models/ApplicantsModel.php (hunk 4 above)
Comment (vincedefreitas) [reply src/Models/ApplicantsModel.php]: Done (Removed)

File: templates/editApplicant.phtml
Comment (Liamslost) [reply templates/editApplicant.phtml:1]: done

Comment (Liamslost) [reply templates/editApplicant.phtml:1]: done apart from being clever

File: src/Controllers/Controller.php (hunk 2 above)
Comment (Liamslost) [reply src/Controllers/Controller.php]: done

Comment (vincedefreitas) [reply src/Controllers/Controller.php]: Left as is, as the controllers required changing anyway

File: src/Models/CohortsModel.php (hunk 9 above)
Comment (fennel3) [reply src/Models/CohortsModel.php]: done!

File: src/Controllers/FormActions/EditApplicantActionController.php (hunk 8 above)
Comment (vincedefreitas) [reply src/Controllers/FormActions/EditApplicantActionController.php]: Validation done by @DBaldwin8 & @fennel3

File: templates/applicants.phtml (hunk 1 above)
Comment (FredJRice) [reply templates/applicants.phtml]: Fun removed.  Buttons now fallen into line.

File: src/Models/ApplicantsModel.php (hunk 3 above)
Comment (Liamslost) [reply src/Models/ApplicantsModel.php]: done

File: public/css/style.css
Comment (FredJRice) [reply public/css/style.css:1]: This is now done.

File: public/css/style.css [hunk 10]
```diff
@@ -1,270 +1,340 @@
...
+--dark-grey: #333333;
+--light-blue: #337ab7;
+--red: #FF0000;
+--io-pink: #FF3366;
+--io-purple: #743DFB;
+--io-purple-dark: #0F0028;
+--io-green: #5FFFB4;
+
+--small: 768px;
 }
```
Comment (mallowmew) [inline public/css/style.css:18]: It is good to see that you have added the variables, but then you don't use them 😔.

For example, you should replace every occurrence of `#dddddd` in this file with `var(--light-grey)`

File: public/css/style.css [hunk 11]
```diff
@@ -1,270 +1,340 @@
...
 
 .btn-width{
-  width: 140px;
+    width: 140px;
 }
 
 .btn:hover {
-  background-color: rgb(0, 248, 131.75);
+    background-color: rgb(0, 248, 131.75);
```
Comment (mallowmew) [inline public/css/style.css]: Isn't this the variable `--io-green`?

File: public/css/style.css [hunk 12]
```diff
@@ -1,270 +1,340 @@
...
-  }
+.archived {
+    color: #FF0000;
+    font-size: 2rem;
+}
 
-  .admin .links {
-    grid-template-columns: 1fr;
-  }
+@media screen and (max-width: 768px) {
```
Comment (mallowmew) [inline public/css/style.css]:
```suggestion
@media screen and (max-width: var(--small)) {
```

File: src/Models/CohortsModel.php [hunk 13]
```diff
@@ -53,4 +53,10 @@ public function getAll(): array
 
         return $cohorts;
     }
+
+    public function getDate (){
+        $query = $this->db->prepare("SELECT `id`, `date` FROM `cohorts`;");
+        $query->execute();
+        return $query->fetchAll();
+    }
```
Comment (mallowmew) [inline src/Models/CohortsModel.php]: This returns multiple dates so it should be `getDates` or `getCohortDates`.

But also why don't you juse use the `getAll` to get a list of all of the cohorts and then access the date from those entities? That's what I meant by saying it looked like this had already been done.

File: templates/applicants.phtml [hunk 14]
```diff
@@ -42,6 +43,7 @@
                     <td><a href="/admin/applicants/<?= $applicant->getId() ?>"><?= $applicant->getName(); ?></a></td>
                     <td><?php echo $applicant->getEmail(); ?></td>
                     <td><?php echo $applicant->getFormattedApplicationDate(); ?></td>
+                    <td><a href="/admin/applicant/edit/<?= $applicant->getId() ?>" class="edit-button">EDIT</a></td>
```
Comment (mallowmew) [inline templates/applicants.phtml]: It's really easy to make the button consistent with other buttons because there is already a CSS class...

```suggestion
                    <td><a href="/admin/applicant/edit/<?= $applicant->getId() ?>" class="btn">EDIT</a></td>
```

Don't reinvent the wheel.

File: public/css/style.css [hunk 15]
```diff
@@ -1,270 +1,340 @@
...
+    font-size: 14px;
+    position: absolute;
+    background-color: #61fed2;
+    color: black;
+    transition: all 0.1s linear;
+    box-shadow: 0 0 0 0 rgb(0,0,0);
+}
+.archive-button:hover {
+    background-color: #dddddd;
 }
```
Comment (mallowmew) [inline public/css/style.css]: Remove these CSS classes, they're redundant / not part of this story.

File: src/Models/ApplicationModel.php [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+            `phone` = :phone,
+            `address` = :address,
+            `heard_about_id` = :heard_about_id,
+            `age_confirmation` = :age_confirmation,
+            `newsletter` = :newsletter,
+            `eligible` = :eligible,
+            `terms` = :terms
+            WHERE `applicant_id` = :id");
+        $query->execute([
+            'id' => $details['id'],
```
Comment (mallowmew) [inline src/Models/ApplicationModel.php]: `id` usages should be applicant ID to disambiguate for later readers whether you mean the applicant ID or the application ID.

**Review by mallowmew — CHANGES_REQUESTED** [review]

File: src/Models/CohortsModel.php (hunk 13 above)
Comment (vincedefreitas) [reply src/Models/CohortsModel.php]: We've done this now :)

File: public/css/style.css (hunk 15 above)
Comment (fennel3) [reply public/css/style.css:340]: ill do this

File: src/Models/ApplicationModel.php (hunk 16 above)
Comment (vincedefreitas) [reply src/Models/ApplicationModel.php]: Done :)

File: public/css/style.css (hunk 10 above)
Comment (vincedefreitas) [reply public/css/style.css:18]: Done

File: public/css/style.css (hunk 11 above)
Comment (vincedefreitas) [reply public/css/style.css]: Done

File: public/css/style.css (hunk 12 above)
Comment (vincedefreitas) [reply public/css/style.css]: Done

File: public/css/style.css (hunk 15 above)
Comment (vincedefreitas) [reply public/css/style.css]: Done

**Review by mallowmew — APPROVED** [review]
